### DISCUSSION TITLE
Vision for project features

### DISCUSSION THREAD

**Discussion by alexanderson1993** [Thorium-Sim/thorium-nova#21] [Ideas]
This is an abridged list of my notes about various projects I want to work on for Thorium and Thorium Nova. Instead of creating a whole bunch of issues and discussions for each of these, I'm going to outline a bunch of stuff here and have it up for discussion. Make sure you identify what specific things you're talking about when you add comments. We can spin these out into separate discussions if folks have lots of thoughts.

Also, feel free to suggest your own ideas for the vision of the project, either here or in a separate discussion.

### Foundation

- Game Engine
  - Everything is simulated. Objects in 3D space, people moving on the player's ship, how power and inventory moves through the ship
  - Networking engine should be robust enough for internet play.
- Release and deployment system
- Star Map Simulation
- Internal Ship Simulation
  - Marauder’s map & other moving entities (cargo containers)
  - Power management & other utilities
- Flight, Station, and Card management - Composable Controls
  - much like Thorium Classic, where there are Stations and Cards (implying multiple screens a user can switch between for their individual job) that can be reconfigured to create a variety of crew stations for any size of crew. Not like Artemis, Empty Epsilon, or Starship Horizons where there are a fixed number of stations that are single-screens. (Although there's nothing wrong with creating screens that include many controls, especially for single-player flights)
- Multiple bridges in the same simulation 
- Plugin System (See Below)
- UI Component Library
  - Elements like buttons, dropdowns, input fields, etc.
  - User interface sound effects

### Cards

- Replicating as many of the cards in Thorium Classic as possible, while adapting them to the broader simulation (Star Map and Marauder’s Map, etc).
- Flight Director Station
  - for adjusting the simulation during the game, playing the role of NPCs, and running through the storyline.
  - The goal is to keep it simple, easy to learn, and easy to use. Complicated Flight Director controls are what turned many people away from Thorium. Thorium Classic is a great example of what not to do.

### Plugin System

- For sharing assets and configuration only, not for modding or extending the core.
- Dependency management is a requirement. Being able to have some plugins depend on other plugins.
- UI Themeing for the Station layout and UI components

### Viewscreen

- **Question**: Should the main view always be the star map, or should it be able to show other things, like videos? How much like Thorium Classic?
- **Question**: What kinds of overlays will be included? (HUD-style)

### Timeline

- A positively delightfully designed authoring experience.
- The ability to target individual entities mid-flight. Perhaps using tags to select any entities that have a given tag.
- Full automation, with the ability to replace the Flight Director. However, it should be designed such that the Flight Director can step in and take over if necessary.
- A bot of some kind to serve as the main computer, or even NPCs

### Training/Tutorial and Documentation system

- Training videos would be created.
  Also have some kind of in-game text training.
- Think about how video games do training. Either they are so simple no training is needed (Pac Man) or they provide contextual popups when appropriate (Zelda BOTW, Skyrim, many others.)

### Tasks & Automated Task Verification

- The ability for the system to keep track of things that the crew needs to do at any given moment, and to recognize when that task is accomplished.
- Likely a crucial piece of the automated Flight Director
- The tasks system in Thorium was a good start, but leaves a lot to be desired. It should be redesigned to better work with the automated timeline and task configs should be included in built-in plugins.

### Single-player Mode

- Single-person cockpit
- This extends nicely to fighter control, where someone in a larger crew can take over an individual fighter.
- Real-time Strategy mode, where you order around a bunch of other ships?

### Show control

- DMX
  - A big thing is figuring out how to do DMX effects, animations, and transitions.
- Sound Effects
- Music Player
- WebRTC sound system
  - Necessary for making sure people get the appropriate sound effects over the internet.
  - Developed as an [SFU](https://webrtcglossary.com/sfu/) (Via https://mediasoup.org) or [MCU](https://webrtcglossary.com/mcu/)
  - Voice chat between players
  - Integrated sound effects system
  - Voice changer for the flight director (https://vox.fyreworks.us)

## Art & Assets

- Innovative card designs
- Graphics for cards
  - Ex. Icons for the power management screen
  - Icons to represent different ship systems
- Graphics for the Lore and Universe
  - As needed by Story writers
  - Faction logos
- Music
- Specialized sound effects
- Creating high-quality, fully featured plugins
  - Ships, planets, images for ship systems

## Story & Mission Writing

- Choose your own adventures
  - The choices you make earlier in the campaign change how things go later in the campaign
  - Choices made earlier should have long-reaching effects on future missions.
  - Flow chart of missions that become unlocked after completing certain missions, and blocking off other possible missions.
  - Many small one-off missions, Gloomhaven style
  - [Gloomhaven Scenario Book](https://online.flippingbook.com/view/145446/8/)

## Extra Projects

<h3 id="thorium-accounts">Thorium Accounts</h3>

- Track Flight Director flights
- Stats about flights flown
- Dashboard for items sold and views
- Reviews and comments for emporium items
- Track purchases
- Feature flagging for beta users?
- Forum/Community
- Subdomain leasing
- Social network aspect - “My Crew”
- Sign into Thorium using email and password or QR code scanning from the mobile website.
- Creating social cards for Facebook and Twitter that have stats about the flight.
- Track some statistics for literally every card (more than one, if necessary). Include all of those statistics on the profile page.
- Track playtime.
- Personalized skins for individual stations. Tactical could have a different skin than Flight.

### Issue Tracker/Bug Reporter

- Allow folks to upvote Thorium Nova tasks
- Integration with Github

### Plugin Store

- Upload their own plugins
- Set a price or release for free
- Attach a payment processor (perhaps Stripe Connect)
- Manage downloads, track analytics for packages

### Discord Bot

- Create a bridge channel
- Request to join a bridge
- Random quote (from Thorium’s quote library)
- Create a private bridge channel
- Verify that people have read the Code of Conduct
- Count the number of messages sent on Discord
- Invite to a role using a special link
- Add certain roles (Notification roles, station roles [helm, engineer, etc])
- Music player (Thorium Nova licensed music)
- Moderation Tools
- Perks for people who boost
- “Issue Report” channel that only allows slash commands for submitting new issue reports or getting the status of an existing issue.

### [Build out Vox more](https://vox.fyreworks.us)

- Add more effects
- Figure out how to do real-time formant shifting
- Integrate it with Thorium Nova directly

**Comment by Xansta** [comment]
**Thorium Nova Core**
**Foundation**
**Flight, Station, and Card management - Composable Controls**
Empty Epsilon is a bit more flexible than Artemis when it comes to consoles. There's the "standard" 5: Helm, Weapons, Engineering, Science, Relay listed as the 5/6 set (the 6th being Captain). There's a condensed set of consoles: Tactical (essentially Helm and Weapons), Operations (essentially Science and Relay) and Engineering+ (Engineering plus shield control) listed as the 3/4 set (the 4th being the Captain). There's also a single pilot console which the developer set up primarily for testing purposes, but which gets used on occasion by players. Then there are a number of subsets: Damage Control (a subset of Engineering), Power Management (another subset of Engineering), Database (a subset of Science), Ship's Log (a subset or Relay), Strategic Map (another subset of Relay). There are also selections that don't have any influence on the ship at all, but are there for auxiliary monitors: ship window (set the view angle), cinematic view (sort of a movie version of whatever action is taking place - an algorithm picks camera views), top down cinematic view.

The EE developer has talked in the past about opening up customization parameters to allow users to construct their own consoles from various functional parts, but it has not happened yet.

I definitely like the flexibility that Thorium offers in terms of being able to construct your own console out of various parts. This allows you to set up your ship for a very wide range of potential crew/players. This works well when you have an event that is planned well in advance and the number of participants is known. You can plan your set and your mission around your planned visitors. It's not so good when you want to put together an ad hoc game. Part of the SBS experience is working together as a crew on a mission, so I tend to play down the consoles with many functions on them. I think a console with either a dedicated set of functions or a limited number of functions fosters communication between crew and helps build the team. Even so, We may want to consider having some kind of semi-standard set of consoles (groups of cards) to make it easy for pickup games.

In an ideal world, the system would ask how many crew were participating and then dynamically configure a number of stations to match that number, putting together various cards as needed or breaking apart consoles as needed depending on how the algorithm was written. Also, in an ideal world, the people who say they are going to participate will actually show up and the people who forgot to RSVP won't show up expecting to participate.

**Comment by alexanderson1993** [comment]
<img width="675" alt="Screen Shot 2021-10-07 at 7 11 42 AM" src="https://user-images.githubusercontent.com/6558157/136373341-9253ac2a-6d42-4c8b-9994-fe62dc9379a4.png">

Good news! Thorium Nova Prime (the old version) has just this very thing. You put in the number of crew and what class of ship you want and it auto-generates the station complement for that flight - based on the capabilities of the ship and the crew count. 

I think we can use this same concept moving forward, since it really does help with making it fast and easy to get into a flight.
